File: macos/BlindfoldCore/Tests/BlindfoldCoreTests/UnprotectedModeMenuTests.swift
import Testing
@testable import BlindfoldCore

/// The Unprotected-mode submenu's activation rows (issue #187 / ADR-0038): the
/// five bounds an operator can pick, always present regardless of whether the
/// mode is currently active.
@Test func itemsOffersTheFiveActivationBoundsWhenNotActive() {
    let items = UnprotectedModeMenu.items(alarm: nil)
    #expect(items == [
        UnprotectedModeMenuItem(label: "Next request only", action: .activate(bound: "next-request", minutes: nil), keyboardShortcut: nil),
        UnprotectedModeMenuItem(label: "For 5 minutes", action: .activate(bound: "timed", minutes: 5), keyboardShortcut: nil),
        UnprotectedModeMenuItem(label: "For 15 minutes", action: .activate(bound: "timed", minutes: 15), keyboardShortcut: nil),
        UnprotectedModeMenuItem(label: "For 30 minutes", action: .activate(bound: "timed", minutes: 30), keyboardShortcut: nil),
        UnprotectedModeMenuItem(label: "Infinite", action: .activate(bound: "infinite", minutes: nil), keyboardShortcut: nil),
    ])
}

/// "Resume protection now" (⌘⇧P, per the issue brief) only appears once the
/// mode is actually active -- appended after the five activation rows, not
/// replacing them, since the operator can still re-pick a different bound.
@Test func itemsAppendsResumeRowWithShortcutWhenAlarmIsActive() {
    let alarm = UnprotectedAlarm(bound: "infinite", remainingSeconds: nil)
    let items = UnprotectedModeMenu.items(alarm: alarm)
    #expect(items.last == UnprotectedModeMenuItem(label: "Resume protection now", action: .resume, keyboardShortcut: "⌘⇧P"))
    #expect(items.count == UnprotectedModeMenu.activationItems.count + 1)
}

/// The submenu is absent entirely (not just disabled) until #188's capability
/// flag is on -- "a fresh install cannot have protection disabled by a rogue
/// local process one POST away" (ADR-0038) applies just as much to the menu
/// surfacing the button as to the endpoint accepting the call.
@Test func isVisibleIsFalseWhenCapabilityIsDisabled() {
    #expect(UnprotectedModeMenu.isVisible(capabilityEnabled: false) == false)
}

@Test func isVisibleIsTrueWhenCapabilityIsEnabled() {
    #expect(UnprotectedModeMenu.isVisible(capabilityEnabled: true) == true)
}

/// A recorded double at #180's control-endpoint boundary (leak-audit's own
/// seam-stub pattern) -- `UnprotectedModeMenu.perform` only ever asserts
/// through this seam, never against the endpoint's own implementation.
private final class RecordingControl: UnprotectedModeControlling, @unchecked Sendable {
    var activateCalls: [(bound: String, minutes: Int?)] = []
    var resumeCallCount = 0

    func activate(bound: String, minutes: Int?) {
        activateCalls.append((bound: bound, minutes: minutes))
    }

    func resume() {
        resumeCallCount += 1
    }
}

/// Each activation row calls #180's control endpoint (`POST
/// /v1/unprotected-mode`) with its bound/minutes, through the seam.
@Test func performActivateCallsControlActivateWithBoundAndMinutes() {
    let control = RecordingControl()
    UnprotectedModeMenu.perform(.activate(bound: "timed", minutes: 15), control: control)
    #expect(control.activateCalls.count == 1)
    #expect(control.activateCalls.first?.bound == "timed")
    #expect(control.activateCalls.first?.minutes == 15)
}

/// "Resume protection now" calls #180's `DELETE /v1/unprotected-mode` through
/// the same seam.
@Test func performResumeCallsControlResume() {
    let control = RecordingControl()
    UnprotectedModeMenu.perform(.resume, control: control)
    #expect(control.resumeCallCount == 1)
}

/// Auto-revert notification (issue #187 / ADR-0038): "enabling is already loud
/// via the icon + the audit event on the proxy side," so this fires only when
/// the alarm drops on its own -- the next-request/timed expiry -- never when the
/// operator themself just clicked "Resume protection now" (they already know).
@Test func shouldNotifyAutoRevertWhenAlarmWasActiveAndIsNowGoneWithoutAManualResume() {
    let previous = UnprotectedAlarm(bound: "timed", remainingSeconds: 1)
    #expect(UnprotectedModeMenu.shouldNotifyAutoRevert(previousAlarm: previous, currentAlarm: nil, manualResumeRequested: false) == true)
}

@Test func shouldNotNotifyAutoRevertWhenTheOperatorJustClickedResume() {
    let previous = UnprotectedAlarm(bound: "infinite", remainingSeconds: nil)
    #expect(UnprotectedModeMenu.shouldNotifyAutoRevert(previousAlarm: previous, currentAlarm: nil, manualResumeRequested: true) == false)
}

@Test func shouldNotNotifyAutoRevertWhenTheAlarmWasNeverActive() {
    #expect(UnprotectedModeMenu.shouldNotifyAutoRevert(previousAlarm: nil, currentAlarm: nil, manualResumeRequested: false) == false)
}

@Test func shouldNotNotifyAutoRevertWhenTheAlarmIsStillActive() {
    let alarm = UnprotectedAlarm(bound: "timed", remainingSeconds: 30)
    #expect(UnprotectedModeMenu.shouldNotifyAutoRevert(previousAlarm: alarm, currentAlarm: alarm, manualResumeRequested: false) == false)
}

@Test func autoRevertNotificationMessageIsThePassThroughEndedCopy() {
    #expect(UnprotectedModeMenu.autoRevertNotificationMessage == "Pass-through ended — full protection restored.")
}

File: windows/Blindfold.Core.Tests/UnprotectedModeMenuTests.cs
using Blindfold.Core;
using Xunit;

namespace Blindfold.Core.Tests;

/// <summary>
/// The Unprotected-mode submenu (issue #197, ADR-0038/0041) -- the C# port of macOS's
/// <c>UnprotectedModeMenu</c> (issue #187/#188). Widened past the Swift original: issue #197's
/// AC explicitly requires the submenu itself to drive the capability toggle (there is no
/// separate Settings surface reachable from the tray yet), so <c>Items</c> takes the place of
/// Swift's <c>isVisible</c> gate -- capability off renders a single enable row instead of the
/// five-bound submenu being absent outright.
/// </summary>
public class UnprotectedModeMenuTests
{
    /// <summary>Fail-closed default: capability disabled shows one row to opt in, never the
    /// five-bound submenu (ADR-0038's "not one loopback POST away").</summary>
    [Fact]
    public void ItemsIsJustTheEnableRowWhenCapabilityIsDisabled()
    {
        var items = UnprotectedModeMenu.Items(capabilityEnabled: false, alarm: null);

        Assert.Equal(
            new[] { new UnprotectedModeMenuItem("Enable Unprotected Mode…", UnprotectedModeAction.EnableCapability, null) },
            items);
    }

    /// <summary>The five activation bounds, always present once the capability is on,
    /// regardless of whether the mode is currently active.</summary>
    [Fact]
    public void ItemsOffersTheFiveActivationBoundsWhenCapabilityIsEnabledAndNotActive()
    {
        var items = UnprotectedModeMenu.Items(capabilityEnabled: true, alarm: null);

        Assert.Equal(
            new[]
            {
                new UnprotectedModeMenuItem("Next request only", UnprotectedModeAction.Activate("next-request", null), null),
                new UnprotectedModeMenuItem("For 5 minutes", UnprotectedModeAction.Activate("timed", 5), null),
                new UnprotectedModeMenuItem("For 15 minutes", UnprotectedModeAction.Activate("timed", 15), null),
                new UnprotectedModeMenuItem("For 30 minutes", UnprotectedModeAction.Activate("timed", 30), null),
                new UnprotectedModeMenuItem("Infinite", UnprotectedModeAction.Activate("infinite", null), null),
                new UnprotectedModeMenuItem("Disable Unprotected Mode…", UnprotectedModeAction.DisableCapability, null),
            },
            items);
    }

    /// <summary>"Resume protection now" only appears once the mode is actually active --
    /// appended after the five activation rows and before the capability-disable row, since
    /// the operator can still re-pick a different bound.</summary>
    [Fact]
    public void ItemsAppendsResumeRowWithShortcutWhenAlarmIsActive()
    {
        var alarm = new UnprotectedAlarm("infinite", null);

        var items = UnprotectedModeMenu.Items(capabilityEnabled: true, alarm: alarm);

        Assert.Equal(new UnprotectedModeMenuItem("Resume protection now", UnprotectedModeAction.Resume, "Ctrl+Shift+P"), items[5]);
        Assert.Equal(UnprotectedModeAction.DisableCapability, items[6].Action);
        Assert.Equal(7, items.Count);
    }

    private sealed class RecordingControl : IUnprotectedModeControlling
    {
        public List<(string Bound, int? Minutes)> ActivateCalls { get; } = new();
        public int ResumeCallCount { get; private set; }
        public List<bool> CapabilityCalls { get; } = new();

        public void Activate(string bound, int? minutes) => ActivateCalls.Add((bound, minutes));
        public void Resume() => ResumeCallCount++;
        public void SetCapability(bool enabled) => CapabilityCalls.Add(enabled);
    }

    [Fact]
    public void PerformActivateCallsControlActivateWithBoundAndMinutes()
    {
        var control = new RecordingControl();

        UnprotectedModeMenu.Perform(UnprotectedModeAction.Activate("timed", 15), control);

        Assert.Equal(new[] { ("timed", (int?)15) }, control.ActivateCalls);
    }

    [Fact]
    public void PerformResumeCallsControlResume()
    {
        var control = new RecordingControl();

        UnprotectedModeMenu.Perform(UnprotectedModeAction.Resume, control);

        Assert.Equal(1, control.ResumeCallCount);
    }

    [Fact]
    public void PerformEnableCapabilityCallsControlSetCapabilityTrue()
    {
        var control = new RecordingControl();

        UnprotectedModeMenu.Perform(UnprotectedModeAction.EnableCapability, control);

        Assert.Equal(new[] { true }, control.CapabilityCalls);
    }

    [Fact]
    public void PerformDisableCapabilityCallsControlSetCapabilityFalse()
    {
        var control = new RecordingControl();

        UnprotectedModeMenu.Perform(UnprotectedModeAction.DisableCapability, control);

        Assert.Equal(new[] { false }, control.CapabilityCalls);
    }

    /// <summary>Auto-revert notification: enabling is already loud via the icon + the proxy-side
    /// audit event, so this fires only when the alarm drops on its own, never when the operator
    /// themself just clicked "Resume protection now".</summary>
    [Fact]
    public void ShouldNotifyAutoRevertWhenAlarmWasActiveAndIsNowGoneWithoutAManualResume()
    {
        var previous = new UnprotectedAlarm("timed", 1);

        Assert.True(UnprotectedModeMenu.ShouldNotifyAutoRevert(previous, null, manualResumeRequested: false));
    }

    [Fact]
    public void ShouldNotNotifyAutoRevertWhenTheOperatorJustClickedResume()
    {
        var previous = new UnprotectedAlarm("infinite", null);

        Assert.False(UnprotectedModeMenu.ShouldNotifyAutoRevert(previous, null, manualResumeRequested: true));
    }

    [Fact]
    public void ShouldNotNotifyAutoRevertWhenTheAlarmWasNeverActive()
    {
        Assert.False(UnprotectedModeMenu.ShouldNotifyAutoRevert(null, null, manualResumeRequested: false));
    }

    [Fact]
    public void ShouldNotNotifyAutoRevertWhenTheAlarmIsStillActive()
    {
        var alarm = new UnprotectedAlarm("timed", 30);

        Assert.False(UnprotectedModeMenu.ShouldNotifyAutoRevert(alarm, alarm, manualResumeRequested: false));
    }

    [Fact]
    public void AutoRevertNotificationMessageIsThePassThroughEndedCopy()
    {
        Assert.Equal("Pass-through ended — full protection restored.", UnprotectedModeMenu.AutoRevertNotificationMessage);
    }
}
